### DISCUSSION TITLE
Sub-beat chord durations

### DISCUSSION THREAD

**Discussion by phatmann** [no-chris/chord-mark#500] [Ideas]
Many songs have passing chords that have a duration of an eighth-note. For example, Elton John's "Don't Let the Sun Go Down On Me". For ChordMark to support that, we would need a way to notate smaller units than a beat in the lyrics and the chords. Has this been considered?

**Comment by phatmann** [comment]


**Reply by no-chris** [reply]
I've been thinking about this a lot. That would also allow transcribing Let it be ;-)

I see three options but I'd like to explore more possibilities, please add more ideas!

**1/ Nashville number system style**

`<Am <G7`

Does not really solves the issue, allows only to denote syncopes. Here the Am would start on the 8th beat immediatly before the bar.

**2/ sub-beat duration marker**

` F.  C/E, Dm7, C..`

is how you'd write Let it be

**3/  Glue character for 8th notes**

` F.  C/E-Dm7 C..`
or ` F.  C/E_Dm7 C..`
or ` F.  C/E:Dm7 C..`

**Reply by phatmann** [reply]
The comma seems like a good choice for an eighth-note since is looks like a "flag". However, we need an extensible notation for sixteenth notes, etc. Continuing in the vein of copying musical notation, it would be natural to use a double, triple, etc. flag. Even though sixteenth note chords are unusual, we still need to handle this, and if we add melody line support it becomes essential.

A semi-colon looks an upside down eighth note. Maybe use that for eighth note, and add extra commas as needed for sixteenth notes, etc.

`F. C/E; Dm7;, C..`

**Reply by no-chris** [reply]
mmmm, I'm not sure I would want to go to the point of describing melody lines or we'll end up replicating [ABC](https://abcnotation.com/) or [Lilypond](https://lilypond.org/text-input.html) notation formats. I'd rather add support or partial support for those.

16th notes in chords duration are unusual to say the least. Do you have some examples? Those might be better handled by some kind of exceptional notation that would opt out the usual chord mark parsing. If we take Lilypond/ABC as an example, both specify the actual duration with a number:
`c16` for Lilypond
`c/16` for ABC

When designing ChordMark I deliberately chose not to go that route because it made chord charts confusing to read, knowing that supporting only 4th notes and above would eventually show limits.

Also encoding is one thing, but we need to keep in mind the rendering as well. Another ChordMark design principle is that it should stay renderable with text only. So how to denote 8th notes in a readable manner is not a trivial issue.

A Chord chart is the most simplified music notation format that I know of. ChordMark re-introduces a bit of complexity with the chord durations but I think the simplification principle should still prevail, even at the cost of not being able to describe everything.

**Reply by phatmann** [reply]
I just deleted the comment about melody lines, as I realized it makes no sense with chord -- and not beat -- markers in the lyrics.

Your point about rendering is a very good one. The comma is natural to _enter_, but when _rendered_ does not seem as naturally understandable as your "glue character" proposal. So let's explore using a glue character.

As for which glue character, the "-" is problematic since it is an alias for minor in chords. The underscore looks good, but since it is used as the lyrical chord marker, it might create some confusion. The colon looks okay, but its normal semantics as a label prefix make it seem too strong to serve as glue.

What do you think about a semi-colon glue character:

`F. C/E;Dm7 C..`

It has the right semantics and looks nice. 

As for sixteenth etc notes, let's establish the eighth-note notation first and then see where that leads us. For example, a repeated glue character becomes a possibility.

**Reply by phatmann** [reply]
While we are opening a can of worms, what about triplets? I have definitely seen triplet passing chords in some jazz songs. 

My proposal is to use brackets for grouping chords into a single beat:

`F. [C/E Dm7 Dmaj7] C..`

Pretty clear I think! And it provides yet another idea for eighth notes:

`F. [C/E Dm7] C..`

If we need a rest, a tilde would work nicely:

`F. [C/E ~ Dmaj7] C..`

I like the grouping idea since it is easy to enter and renders nicely.

**Reply by no-chris** [reply]
>The comma is natural to enter

Then maybe we should keep it as an input? It doesn't have to be _rendered_ as a coma, but it fits ChordMark philosophy of being easy to write, whereas with brackets we would need to enter 2 characters for 8th notes

`F. C/E, Dm7, C..` would be rendered `F. [C/E Dm7] C..`

I have mixed feelings regarding the brackets, because I had in mind the idea of allowing optional chords, like variations that are sometimes used in jazz charts

`A7 (A#°) Bm`

Still I feel they are a viable option, we could even add a `.` after the brackets to make it even more explicit what is happening

`F. [C/E Dm7]. C..`

I don't know that many songs with 8th chords, but the topic seems to be popular in [BandInABox forums](https://www.pgmusic.com/forums/ubbthreads.php?ubb=showflat&Number=113034), so the demand is clearly there.

>  let's establish the eighth-note notation first and then see where that leads us.

I fully agree. Let's start with this and see later for triplets, 16th notes, and so forth.

Maybe we could try to compile a list of songs that use 8th note to try out the syntax and make sure it fits? 
I also have the vague feeling that a notation for syncopes could be useful as well.
BandInABox has this `push` concept with carets `^` or double `^^`
https://www.pgmusic.com/manuals/bbw2022full/chapter6.htm#chord-entry

**Reply by phatmann** [reply]
Reviewing this thread so far, here are my comments:
* A comma can look very similar to a period, especially with a small font. It is also frequently mistyped instead of a period. Might be worth avoiding it.
* As you said, the most frequently used eighth note chords are the "push" chords going into the next verse. Having a shorthand notation for this makes sense. The Nashville "<" notation you propose seems very suitable.
* For all other sub-beat use cases, we need to be able to handle them for completeness, but they don't need to be quick to enter, as they are uncommon. My grouping idea seems suitable for this.
* If we need to preserve brackets ("[]") for other uses, we can use a different delimiter for groups; e.g. braces ("{}")

Based on this, I make the following proposal:

### Push chords

For chords that are played on the final eighth note of a measure, the chord Is preceded by a "<":

`A <A/C# D`

### Groups

Chords with a duration less than one beat are expressed using a group notation. A group is delimited using braces:

`A {B B/D#} E`

A group is one beat long. The number of elements in the braces determines the length of the group. That is, 2 elements means eighth notes, 3 means triplets, 4 sixteenth notes etc.

A chord can be followed by one or more periods to extend it into additional sub-beats:

`A {B . A/C# B/D#}`

The above would be a 1/16 note group, with the B as an 1/8 note.

**Reply by no-chris** [reply]
Nice! 👏🏻 I think we are getting to something very cool

**Reply by phatmann** [reply]
Glad you like it! What's the next step? Should I make a PR, at least for the docs?

**Reply by no-chris** [reply]
Let's sleep with it for a while and experiment with a few transcriptions 

```
C.. G.. Am. {Am Am/G} F..
When I _find myself in _times of trouble, _Mother _Ma_ry _comes to me
C.. G.. F. {C/E Dm7} C..
_Speaking words of _wisdom, let it _be
```
=>
```
       |C             G                 |Am.   (Am Am/G). F.. |
When I find myself in times of trouble, Mother Ma  ry     comes to me
|C                G              |F. (C/E Dm7). C.. |
Speaking words of wisdom, let it be
```
=>
```
|C   G   |Am.   (Am Am/G).   F..   |
|C   G   |F.    (C/E Dm7).   C..   |
```

**Reply by no-chris** [reply]
```
C.. G/B.. Am.. <F..
_ No _woman no cr_y
C.. (F C/E) (Dm7 C) C.. G..
_ No _wo_man _no_cry 
```
=>
```
|C    G/B       |Am   <F |
   No woman no cry
|C..    (F C/E). (Dm7 C).  |C   G  |
     No wo man   no   cry
```
=>
```
|C     G/B                |Am   <F  |
|C..   (F C/E). (Dm7 C).  |C    G   |
```

**Reply by phatmann** [reply]
Looks cool so far. My comments:
* Since chord symbols can have parentheses in them, don't we need a different delimiter?
* Since groups are always one beat, perhaps the period after the group is not needed.
* To solve both issues, how about a colon as delimiter, for both editing and rendering. The period is "built in" 😄 

`C.. :F C/E:`

**Reply by phatmann** [reply]
Any thoughts on this?

**Comment by no-chris** [comment]
Sorry I missed that message somehow!

- Parentheses: yes, possibly we need another symbol. I'd like to explore rendering options maybe with slightly different colors or even nothing, but I like the readability of the parentheses vs the brackets/square brackets
- I like the consistency but I agree, that needs to be experimented a bit
- tbh I find this one a bit confusing to read 😁

**Comment by phatmann** [comment]
@no-chris Any decisions about the sub-beat chord format?

**Reply by no-chris** [reply]
I think we have the basics, now it's a matter of implementing it and fine tuning the syntax!

**Comment by no-chris** [comment]
This has been implemented in https://github.com/no-chris/chord-mark/pull/587

It is part of ChordMark 0.13.0 beta version: https://github.com/no-chris/chord-mark/releases/tag/v0.13.0-beta.0
The beta version of Chord Chart Studio using this is available here: https://deploy-preview-610--chord-chart-studio.netlify.app/app

**Comment by phatmann** [comment]
Hi Chris.

I am super excited to see this in action, but when I try the sub-beat chords the rendering is not working. Maybe the code did not deploy?

**Reply by no-chris** [reply]
https://deploy-preview-610--chord-chart-studio.netlify.app/app

with:

```
#i
C.. G.. 
Am. {Am Am/G} FM7. F6.
C.. G.. 
F. {C/E Dm7} C..

#v
C.. G..
When I _find myself in _times of trouble
Am. {Am Am/G} FM7. F6.
_ Mother _Ma_ry _comes to _me
%%
_Speaking words of _wisdom
F. {C/E Dm7} C..
Let it _be _ _ _

C.. G..
And _in my hour of _darkness
Am. {Am Am/G} FM7. F6.
She is _standing _right _in _front of _me
%%
_Speaking words of _wisdom
F. {C/E Dm7} C. G.
Let it _be _ _ _

#c
Am.. G.. 
Let it _be, let it _be
F.. C..
Let it _be, let it _be
C.. G.. 
_ Whisper words of _wisdom
F. {C/E Dm7} C..
Let it _be

#b x2
F. {C/E Dm7} C. {Bb F/A}
G. F. C.. 
```

**Comment by phatmann** [comment]
Rendering is indeed working:
<img width="703" alt="image" src="https://user-images.githubusercontent.com/36146/212977492-f9a2b377-073b-4c48-b938-0d56a34aec81.png">
So great to see this in action. A few questions:
* Can we get the first chord in the sub-beat group to align with the syllable?
* Do we actually need to render the braces? Since all the other chords have a period next to them, the absence of a period for the subgroup chords may be a sufficient clue.
* Taking this thinking further, if the user inputs "Am. Am Am/G FM7. F6.", the system could recognize that Am/G and FM7 each get eighth-note values without the braces.

**Comment by no-chris** [comment]
Thanks for the feedback!

> Can we get the first chord in the sub-beat group to align with the syllable?

Duh, I was afraid someone would ask for that 😱 I guess we could yes, I did it for the the bar separator, it was painful enough, but I guess replicating it for sub-beats should be easier now that the hard work is done.

> Do we actually need to render the braces? Since all the other chords have a period next to them, the absence of a period for the subgroup chords may be a sufficient clue.

That's easy enough to experiment, we can just disable the rendering of the braces with `printSubBeatDelimiters: false`. It is live if you refresh the deploy preview. I like the clarity provided by the braces, though.

> Taking this thinking further, if the user inputs "Am. Am Am/G FM7. F6.", the system could recognize that Am/G and FM7 each get eighth-note values without the braces.

Nice! I'll keep this idea as a nice to have feature, as it would require rewriting the parsing algorithm, and since sub-beat duration chords are not that common in the end, I think the extra work of writing the braces is acceptable for now.

**Comment by phatmann** [comment]
Ah, a preference is a perfect solution. I like the way it looks with no braces.

I believe the reason I don't like the rendered braces is that they are very "busy" compared with the minimalist vibe of the rendered content. Perhaps a different delimiter would be cleaner? It seems that only "paired" (i.e. different start and end characters) delimiters are clear enough to avoid ambiguity in an environment with flexible spacing. Some ideas for rendering:

⟨ Am Am/G ⟩
‹ Am Am/ G ›
❲ Am Am/G ❳
⊂ Am Am/G ⊃
⊢ Am Am/G ⊣
⋋ Am Am/G ⋌
⫭ Am Am/G ⫬
⟪ Am Am/G ⟫
⁌ Am Am/G ⁍
⨫ Am Am/G ⨬
⨽ Am Am/G ⨼
⧼ Am Am/G ⧽

**Comment by no-chris** [comment]
I agree with the "minimalist vibe". 
I like 
⟨ Am Am/G ⟩
‹ Am Am/ G ›
But I still have in mind the syncope notation (Nashville style) for a later improvement.

I like plain square brackets actually, I feel they are more in line with the rest of the rendering

![Screenshot 2023-01-17 at 22 58 35](https://user-images.githubusercontent.com/11467480/213021794-78c51f1c-28d2-4b09-84a5-85f5d9f46c83.png)

(Note that I've managed to fix the delimiter position 😅 https://github.com/no-chris/chord-mark/pull/597

**Reply by no-chris** [reply]
Although on a second though square brackets do not really stand out from the bar separator `|`...

**Comment by phatmann** [comment]
With the delimiter position correct and brackets it looks much nicer. However, I thought we decided not to use brackets as input since they are sometimes used in chord symbols; am I wrong? In any case I agree that brackets probably do look too much like bars.

**Reply by phatmann** [reply]
At least `‹Am Am/ G›` does not look like bars 😄

**Comment by no-chris** [comment]
Parenthesis are used in chord symbols, not square brackets

**Reply by no-chris** [reply]
reading the discussion above I realise I created this confusion indeed. 
I'll blame it on my limited english vocabulary with regard to punctuation symbols 😅 (braces, brackets, parenthesis, square brackets, etc. 🤯)

**Comment by phatmann** [comment]
Ok, so square brackets `[]`  are available for input as sub-beat delimiters. It probably makes sense to use them instead of braces then.

As for what we render, I still agree with you that square brackets look too much like bars. But the input character can be different than output character.

**Comment by no-chris** [comment]
I've updated the deploy preview with square brackets both for parsing & rendering and adapted the styling so they would stand out more from bar separators. Let me know what you think

```
#i
C.. G..
Am. [Am Am/G] FM7. F6.
C.. G..
F. [C/E Dm7] C..

#v
C.. G..
When I _find myself in _times of trouble
Am. [Am Am/G] FM7. F6.
_ Mother _Ma_ry _comes to _me
%%
_Speaking words of _wisdom
F. [C/E Dm7] C..
Let it _be _ _ _
C.. G..
And _in my hour of _darkness
Am. [Am Am/G] FM7. F6.
She is _standing _right _in _front of _me
%%
_Speaking words of _wisdom
F. [C/E Dm7] C. G.
Let it _be _ _ _

#c
Am.. G..
Let it _be, let it _be
F.. C..
Let it _be, let it _be
C.. G..
_Whisper words of _wisdom
F. [C/E Dm7] C..
Let it _be _ _ _

#b x2
F. [C/E Dm7] C. [Bb F/A]
G. F. C..
```

**Comment by phatmann** [comment]
You made the brackets stand out from the bars, so that is good. But it is becoming clear to me that using delimiters for sub-beats and then postfix punctuation for beats is too dissonant. I like the idea of delimited groups in principal, but in practice it is messing up your minimalist ethos.

So, let's go back to one of our original ideas: the comma.

Am. Am, Am/G, FM7. F6.
_ Mother _Ma_ry _comes to _me

The parser sees two chords in a row with commas, so it knows they are eighth notes.

This renders as:

```
Am.    Am,  Am/G, FM7.      F6.
Mother Ma   ry    comes to  me
```

That is clean and simple, for input and output.

For the pathological case of an eighth note pair next to a triplet, we could still allow optional grouping via tick marks.

A. Am, Am/G, G. \`C, F, C,\`
_This is an _eighth _note  _pair followed by _a _trip_let 

```
A.          Am,    Am/G, G.              `C, F,   C,`
This is an  eighth note  pair followed by a  trip let
```

My original issue with the comma is that it is hard to distinguish from a period when rendered. However, as an input character it seems perfect, with the proper tail of an eighth note!

**Comment by no-chris** [comment]
Thanks for all your feedback! You never seem to run out of ideas 😄 

I think I like the delimiter option better, for the following reasons:
- I find it easier to follow when sightreading, because you get an instant view of how many chords are in the beats. Whereas if you have to identify all the next chords that have a coma and then do the math, I think this is less straightforward to do, especially since the comma and the dot are not that easy to differentiate
- 8th notes are often used over 2 beats for transitions, and then we already need to fallback to the edge case handling
```
C.. [F C/E] [Dm7 C] C.. G..
_ No _wo_man _no_cry 
```
- I'm not found of the idea of having 2 ways to render the same thing, as in your example (one with backticks, one without)

**Reply by phatmann** [reply]
That all makes sense. So then I propose backticks as the delimiter. They have the "tail" aspect of a comma, are lighter than a bracket, and look good for input and output. I know they are not "paired" like brackets, but they are easy to visually parse, so it seems like a good trade off.

**Reply by no-chris** [reply]
So here are both options:

![Screenshot 2023-01-19 at 22 39 42](https://user-images.githubusercontent.com/11467480/213568204-bf3463ad-9ee1-427c-9917-474638cca133.png)
![Screenshot 2023-01-19 at 22 40 39](https://user-images.githubusercontent.com/11467480/213568206-740b1e30-ec1c-4faa-a4a1-97afa36757ef.png)

Personally I still feel the square brackets are a better option.

It would probably help to have more opinions, even though not many people get lost around here. 
Since I'm ChordMark's most active user I might want to leave it to my taste for now 😁  
But once more people voice their opinions I would gladly align with what the majority is asking

**Reply by no-chris** [reply]
In any case, ChordMark also accept custom post-processing filters, so it would be very easy for a client code to configure it

**Reply by phatmann** [reply]
I suppose we just need to standardize on the input characters, and brackets are just fine. It makes sense for your _reference_ implementation to render the same characters in the output, but other implementations have more freedom.

**Reply by no-chris** [reply]
ChordMark's [API](https://github.com/no-chris/chord-mark/blob/a4e1f8d8a2a459da4ae733f937c4f28048ca55fb/packages/chord-mark/src/renderer/components/renderSong.js#L29) has quite a few configuration options - as well as the underlying ChordSymbol library - so yes, other implementations have a lot of flexibility

**Reply by no-chris** [reply]
And thanks again for all the feedback and challenges, I think we have significantly improved the original implementation.
